rgw/lc: add per-bucket lifecycle performance monitoring

mheler opened the pull request:
This PR implements real-time per-bucket lifecycle (LC) monitoring through performance counters exposed via the admin socket. Operators can now track LC processing progress, identify bottlenecks, and monitor processing rates.

Currently, there's limited visibility into lifecycle processing at the per-bucket level. When LC processing is slow or stuck, operators have no way to determine:
  - Which buckets are currently being processed
  - How many objects have been scanned
  - How many objects are pending LC actions
  - How long processing has been running
 
Configuration options introduced with this PR:
  - rgw_lc_counters_cache: Enable/disable feature (default: false)
  - rgw_lc_counters_cache_size: Max cached buckets (default: 10000)
  - rgw_lc_counters_batch_size: Flush threshold (default: 5000)

Introduces 4 new per-bucket lifecycle metrics in the rgw_lc_per_bucket namespace:
  - start_time - When LC processing started (seconds since Unix epoch)
  - end_time - When LC processing finished (seconds since Unix epoch)
  - objects_scanned - Total objects scanned in current run
  - objects_pending - Objects currently queued for LC actions

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [X] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [X] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [X] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

mheler committed:
rgw/lc: add per-bucket lifecycle performance monitoring

Implements real-time per-bucket lifecycle (LC) monitoring via performance
counters exposed through the admin socket.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,75 @@ Cache sizing can depend on a number of factors. These factors include:
+
+Information about lifecycle metrics can be seen in the ``rgw_lc_per_bucket`` sections of the output of the ``counter schema`` command.
+
+To view lifecycle metrics in the Ceph Object Gateway go to the ``rgw_lc_per_bucket`` sections of the output of the ``counter dump`` command::
```
ditto

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,75 @@ Cache sizing can depend on a number of factors. These factors include:
+
+This config value is set in Ceph via the command ``ceph config set client.rgw rgw_lc_counters_cache true``
+
+Since the lifecycle metrics are labeled perf counters, they live in memory. If the Ceph Object Gateway is restarted or crashes, all counters in the Ceph Object Gateway, whether in a cache or not, are lost.
```
Suggest s/If/When/

anthonyeleven replied:
Suggest wrapping the option name in a confval as below.

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,75 @@ Cache sizing can depend on a number of factors. These factors include:
+Lifecycle Counter Cache Size & Eviction
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``rgw_lc_counters_cache_size`` can be used to set number of entries in the cache.
```
s/can/option can/

Also suggest wrapping in :confval:\`rgw_lc_counters_cache_size\`

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -209,6 +278,12 @@ The following Ceph Object Gateway op metrics related settings can be set via ``c
+.. confval:: rgw_lc_counters_cache
+.. confval:: rgw_lc_counters_cache_size
+.. confval:: rgw_lc_counters_batch_size
+
```
Is this redundant given the above?

mheler commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,75 @@ Cache sizing can depend on a number of factors. These factors include:
+
+Lifecycle metrics are labeled counters tracked per-bucket in the ``rgw_lc_per_bucket`` section.
+
+Information about lifecycle metrics can be seen in the ``rgw_lc_per_bucket`` sections of the output of the ``counter schema`` command.
```
Done

mattbenjamin commented on src/rgw/rgw_lc.cc:
```diff
@@ -1522,9 +1601,9 @@ void LCOpRule::update()
+		      optional_yield y, LCBatchCounters* batch_counters)
 {
-  lc_op_ctx ctx(env, o, next_key_name, num_noncurrent, effective_mtime, dpp);
+  lc_op_ctx ctx(env, o, next_key_name, num_noncurrent, effective_mtime, dpp, batch_counters);
```
this one is ok, because dpp is another structure pointer

mheler commented on src/rgw/rgw_perf_counters.cc:
```diff
@@ -232,6 +232,92 @@ CountersManager::~CountersManager() {
+  if (!lc_counters_cache) {
+    return;
+  }
+  std::string key = ceph::perf_counters::key_create(rgw_lc_counters_key, {{"bucket", bucket_name}});
```
I can pass directly as a temporary if that is preferred

mheler replied:
@cbodley Thanks for that recommendation, I made those changes as requested!

LCBatchCounters batches lifecycle counter updates in memory and flushes them periodically to reduce the overhead of frequent PerfCounters cache lookups during object processing.

anthonyeleven commented:
Please select a choice from the other two sections of the checklist, that's required for the first merge check to pass.  The RTD thing is an issue that I think @bluikko has a bit of insight regarding.

mheler commented:
> Please select a choice from the other two sections of the checklist, that's required for the first merge check to pass. The RTD thing is an issue that I think @bluikko has a bit of insight regarding.

Done, let me know if anything else is needed on the docs side.

BBoozmen commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,73 @@ Cache sizing can depend on a number of factors. These factors include:
+
+The :confval:`rgw_lc_counters_cache_size` option can be used to set number of entries in the cache.
+
+Counters are evicted from the cache once the number of counters in the cache are greater than the cache size config variable. When the number of cached counters exceeds this value, the least recently used (LRU) counters are evicted.
```
This comment doesn't request any code/doc change but just to understand the implications of this eviction logic on the premise of this PR:

> This PR implements real-time per-bucket lifecycle (LC) monitoring through performance counters exposed via the admin socket. Operators can now track LC processing progress, identify bottlenecks, and monitor processing rates.

In a highly scaled system (e.g., number of buckets > 100K; most of which have some LC rule configured), won't we lose track of the bucket(s) we'd be interested in?

For example, a bucket's LC processing stalled for some reason and some other buckets are still being processed, depending on how frequently we dump/poll perf, we may not even notice it, right? I mean the record(s) for problematic bucket(s)' are already evicted?

On a similar note, can you please elaborate on how you make use of these new metrics in your clusters? Do you have some sample dashboards that show how these new counters pan out in a scaled cluster?

mheler replied:
@BBoozmen Sorry I misunderstood your question, the cache is persistent until the cache size is met (LRU happens) or RGW daemon is restarted.

Trivially yes, you will want to adjust the cache size based off your environment size. I figured 10k was a good number since that is also used for rgw_user_counters_cache_size, but I am open to setting a higher default if we think that's required.

ceph-jenkins commented:
Can one of the admins verify this patch?

mheler commented:
reopen

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+   * - Name
+     - Type
+     - Description
+   * - start_time
```
suggest double backticks around the names, as they're literals.

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+
+Lifecycle metrics are labeled per-bucket in the ``rgw_lc_per_bucket`` section.
+
+The action metrics (``objects_expired``, ``objects_noncurrent_expired``, ``objects_dm_expired``, ``objects_transitioned``, ``objects_mpu_aborted``) and ``objects_scanned`` are monotonic counters that accumulate across every LC run for the lifetime of the Ceph Object Gateway process. To compute per-run deltas in Prometheus, use ``increase()`` over the LC interval (for example, ``increase(rgw_lc_pe ...
```
nit: please break this into shorter lines.

anthonyeleven replied:
Might a 24h interval capture metrics from two LC runs? Since LC processing runs once a day for N hours. Or maybe with the `start_time` label each run will be presented as a separate series, so this concern is moot?

mheler replied:
I rephrased this so that it makes more sense, and used [24h] as an example that an operator could use.

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+
+Information about lifecycle metrics can be seen in the ``rgw_lc_per_bucket`` section from the output of the ``counter schema`` command.
+
+To view lifecycle metrics in the Ceph Object Gateway go to the ``rgw_lc_per_bucket`` section from the output of the ``counter dump`` command::
```
I think you mean something like

>To retrieve lifecycle metrics from a `radosgw` daemon's admin socket...

mheler replied:
Changed, thanks for that suggestion.

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+Lifecycle Metric Usage Examples
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following examples show how to use each per-bucket lifecycle metric. PromQL examples assume the ceph-exporter is feeding metrics into Prometheus; admin-socket examples use ``ceph daemon`` directly against an RGW.
```
assume that the ``ceph-exporter`` is being scraped by Prometheus.

I'd make the ; a new sentence.

against a specific ``radosgw`` daemon instance

mheler replied:
Done

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+``start_time`` (gauge, Unix epoch seconds)
+""""""""""""""""""""""""""""""""""""""""""
+
+Wall-clock time the most recent LC run started for the bucket. ``0`` means LC has not run for this bucket since the RGW started.
```
The timestamp when the most recent LC run started for the bucket.

mheler replied:
Done

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+
+  .. code-block:: promql
+
+      time() - rgw_lc_per_bucket_start_time < 3600 and rgw_lc_per_bucket_start_time > 0
```
I love that you give lots of PromQL examples.

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+``end_time`` (gauge, Unix epoch seconds)
+""""""""""""""""""""""""""""""""""""""""
+
+Wall-clock time the most recent LC run finished for the bucket. ``0`` while ``start_time > 0`` means LC is currently in progress on that RGW.
```
I think some readers, especially ESOL, might not get "wall-clock" so I'd just go with timestamp.

mheler replied:
Done

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+
+Cumulative noncurrent-version objects expired by LC. Useful only on versioned buckets.
+
+* Per-run noncurrent cleanup volume:
```
I'd write something more specific vs "volume". Is this the number removed?

mheler replied:
I reworded this to be `Noncurrent-version objects removed per run`

anthonyeleven commented on doc/radosgw/metrics.rst:
```diff
@@ -191,6 +191,332 @@ Cache sizing can depend on a number of factors. These factors include:
+        + increase(rgw_lc_per_bucket_objects_mpu_aborted[24h])
+      )
+
+* Bucket-level LC efficiency (fraction of scanned objects that produced an action):
```
Is this really efficiency? Maybe a different term?

mheler replied:
I have reworded it as `Per-bucket LC hit rate (fraction of scanned objects acted on)`

anthonyeleven reviewed (commented):
Variety of docs requests.

mheler committed:
rgw/lc: drop per-bucket LC counters to PRIO_DEBUGONLY

mgr's perf-schema bridge silently drops labeled counters on the way
out, so shipping the per-bucket LC counters up through MgrReport just
costs ingest memory for data mgr can't expose anyway. ceph-exporter
already handles labeled counters via the daemon admin socket, so make
that the only path.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

bluikko commented:
There are several `CRITICAL` complaints from Sphinx build. These usually happen if the level of section titles is not in the same order consistently in the document. For example, first using the order `----`, `~~~~`, `^^^^` and then later have `^^^^` directly after a `----` level section title:

```console
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:273: CRITICAL: Title level inconsistent:
Lifecycle Counter Cache
^^^^^^^^^^^^^^^^^^^^^^^
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:282: CRITICAL: Title level inconsistent:
Lifecycle Counter Cache Size & Eviction
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:289: CRITICAL: Title level inconsistent:
Lifecycle Counter Batching
^^^^^^^^^^^^^^^^^^^^^^^^^^^
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:296: CRITICAL: Title level inconsistent:
Lifecycle Metric Usage Examples
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:306: CRITICAL: Title level inconsistent:
``start_time`` (gauge, Unix epoch seconds)
""""""""""""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:329: CRITICAL: Title level inconsistent:
``end_time`` (gauge, Unix epoch seconds)
""""""""""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:354: CRITICAL: Title level inconsistent:
``objects_scanned`` (counter)
"""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:380: CRITICAL: Title level inconsistent:
``objects_pending`` (gauge)
"""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:404: CRITICAL: Title level inconsistent:
``objects_expired`` (counter)
"""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:428: CRITICAL: Title level inconsistent:
``objects_noncurrent_expired`` (counter)
""""""""""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:447: CRITICAL: Title level inconsistent:
``objects_dm_expired`` (counter)
""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:465: CRITICAL: Title level inconsistent:
``objects_transitioned`` (counter)
""""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:489: CRITICAL: Title level inconsistent:
``objects_mpu_aborted`` (counter)
"""""""""""""""""""""""""""""""""
/home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:506: CRITICAL: Title level inconsistent:
Combined Examples
"""""""""""""""""
```

mheler commented:
> There are several `CRITICAL` complaints from Sphinx build. These usually happen if the level of section titles is not in the same order consistently in the document. For example, first using the order `----`, `~~~~`, `^^^^` and then later have `^^^^` directly after a `----` level section title:
> 
> ```
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:273: CRITICAL: Title level inconsistent:
> Lifecycle Counter Cache
> ^^^^^^^^^^^^^^^^^^^^^^^
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:282: CRITICAL: Title level inconsistent:
> Lifecycle Counter Cache Size & Eviction
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:289: CRITICAL: Title level inconsistent:
> Lifecycle Counter Batching
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:296: CRITICAL: Title level inconsistent:
> Lifecycle Metric Usage Examples
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:306: CRITICAL: Title level inconsistent:
> ``start_time`` (gauge, Unix epoch seconds)
> """"""""""""""""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:329: CRITICAL: Title level inconsistent:
> ``end_time`` (gauge, Unix epoch seconds)
> """"""""""""""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:354: CRITICAL: Title level inconsistent:
> ``objects_scanned`` (counter)
> """""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:380: CRITICAL: Title level inconsistent:
> ``objects_pending`` (gauge)
> """""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:404: CRITICAL: Title level inconsistent:
> ``objects_expired`` (counter)
> """""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:428: CRITICAL: Title level inconsistent:
> ``objects_noncurrent_expired`` (counter)
> """"""""""""""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:447: CRITICAL: Title level inconsistent:
> ``objects_dm_expired`` (counter)
> """"""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:465: CRITICAL: Title level inconsistent:
> ``objects_transitioned`` (counter)
> """"""""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:489: CRITICAL: Title level inconsistent:
> ``objects_mpu_aborted`` (counter)
> """""""""""""""""""""""""""""""""
> /home/docs/checkouts/readthedocs.org/user_builds/ceph/checkouts/66064/doc/radosgw/metrics.rst:506: CRITICAL: Title level inconsistent:
> Combined Examples
> """""""""""""""""
> ```

Thanks, I went ahead and fixed these. Please let me know if you find any other issues.

mattbenjamin commented on src/rgw/rgw_perf_counters.cc:
```diff
@@ -238,7 +238,12 @@ ceph::perf_counters::PerfCountersCache *lc_counters_cache = nullptr;
 void add_lc_counters(PerfCountersBuilder *pcb) {
-  pcb->set_prio_default(PerfCountersBuilder::PRIO_USEFUL);
+  /*
+   * mgr can't expose labeled counters via its prometheus module, so
```
ack, that's more than I knew about this

mattbenjamin reviewed (approved):
looks great, thank you

mattbenjamin commented:
the only question I have is around tenant vs account (@cbodley )

cbodley commented:
> the only question I have is around tenant vs account (@cbodley )

@mattbenjamin if you're referring to `"labels": { "bucket": "mybucket", "tenant": "" }`, we require both a bucket name and its tenant namespace to uniquely identify a bucket. i don't think the owning account or user matters here

anrao19 commented:
pr testing is in progress .. 
https://tracker.ceph.com/issues/76521

anrao19 commented:
pr testing completed, and approved by @ivancich. More details in : https://tracker.ceph.com/issues/76521.
@mheler , if there is no further testing needed pr can be merged
